desktop/src/backends/navigator.rs

```diff
@@ -277,4 +294,145 @@ impl NavigatorBackend for ExternalNavigatorBackend {
         }
         url
     }
+
+    fn connect_socket(
+        &mut self,
+        host: String,
+        port: u16,
+        timeout: Option<Duration>,
+        handle: SocketHandle,
+        receiver: Receiver<Vec<u8>>,
+        sender: Sender<SocketAction>,
+    ) {
+        let addr = format!("{}:{}", host, port);
+        let is_allowed = self.socket_allowed.contains(&addr);
+        let socket_behavior = self.socket_behavior;
+
+        let future = Box::pin(async move {
+            match (is_allowed, socket_behavior) {
+                (false, SocketBehavior::Unrestricted) | (true, _) => {} // the process is allowed to continue. just dont do anything.
+                (false, SocketBehavior::Deny) => {
+                    // Just fail the connection.
+                    sender
+                        .send(SocketAction::Connect(handle, true))
+                        .expect("working channel send");
+
+                    tracing::warn!(
+                        "SWF tried to open a socket, but opening a socket is not allowed"
+                    );
+
+                    return Ok(());
+                }
+                (false, SocketBehavior::Ask) => {
+                    let attempt_sandbox_connect = AsyncMessageDialog::new().set_level(MessageLevel::Warning).set_description(&format!("The current movie is attempting to connect to {:?} (port {}).\n\nTo allow it to do so, click Yes to grant network access to that host.\n\nOtherwise, click No to deny access.", host, port)).set_buttons(MessageButtons::YesNo)
+                    .show()
+                    .await;
+
+                    if !attempt_sandbox_connect {
+                        // fail the connection.
+                        sender
+                            .send(SocketAction::Connect(handle, true))
+                            .expect("working channel send");
+
+                        return Ok(());
+                    }
+                }
+            }
+
+            let host2 = host.clone();
+
+            let mut pending_write = vec![];
+
+            let stream = if let Some(timeout) = timeout {
+                TcpStream::connect((host, port))
+                    .or(async {
+                        Timer::after(timeout).await;
+                        Result::<TcpStream, io::Error>::Err(io::Error::new(ErrorKind::TimedOut, ""))
+                    })
+                    .await
+            } else {
+                TcpStream::connect((host, port)).await
+            };
+
+            let mut stream = match stream {
+                Ok(stream) => {
+                    sender
+                        .send(SocketAction::Connect(handle, true))
+                        .expect("working channel send");
+
+                    stream
+                }
+                Err(err) => {
+                    warn!("Failed to connect to {}:{}, error: {}", host2, port, err);
```

Comment:
AFAIK a timed out connection will throw a SecurityErrorEvent in Flash. It's probably not necessary to implement that right now, but it might be good to add a TODO notice here for that.

Reply:
Is there a specific error message, id?

Reply:
I added a way to differentiate between a failed and timed out connection.